src/ConfigFileParser.cpp

```diff
@@ -0,0 +1,534 @@
+        throw std::runtime_error("More than one root path");
+
+    if (rootPath[rootPath.length() - 1] == '/')
+        rootPath = rootPath.substr(0, rootPath.length() - 1);
+
+    if (block == SERVER)
+        m_configFile.servers[m_configFile.serverIndex].root = rootPath;
+    else if (block == LOCATION)
+        m_configFile.servers[m_configFile.serverIndex].locations[m_configFile.servers[m_configFile.serverIndex].locationIndex].root = rootPath;
+}
+
+/**
+ * @brief Reads and checks the value of the directive in the current line of the config file
+ *
+ * @details This function is called when the directive is valid.
+ *          It calls the appropriate function to read the value of the directive.
+ *          It throws an exception if the value is invalid.
+
+ * @param directive Is the the directive which value is being read and checked
+ */
+void ConfigFileParser::readServerDirectiveValue(const std::string& directive, const std::string& value)
+{
+    if (directive == "listen") {
+        readIpAddress(value);
+        readPort(value);
+    } else if (directive == "root")
+        readRootPath(SERVER, value);
+}
+
+/**
+ * @brief Gets the directive from the line
+ *
+ * @param line String containing the directive and value
+ * @return std::string The extracted directive
+ */
+std::string ConfigFileParser::getDirective(const std::string& line) const
+{
+    size_t firstWhiteSpaceIndex = line.find_first_of(" \t\n\v\f\r");
+    std::string directive = line.substr(0, firstWhiteSpaceIndex);
+
+    removeLeadingAndTrailingSpaces(directive);
+
+    return directive;
+}
+
+/**
+ * @brief Gets the value from the line
+ *
+ * @param line String containing the directive and value;
+ * @return std::string  The extracted value
+ */
+std::string ConfigFileParser::getValue(const std::string& line) const
+{
+    if (getDirective(line) == "location")
+        return line;
+
+    size_t semicolonIndex = line.find(';');
+
+    size_t firstWhiteSpaceIndex = line.find_first_of(" \t\n\v\f\r");
+    std::string value = line.substr(firstWhiteSpaceIndex, semicolonIndex - firstWhiteSpaceIndex);
+
+    removeLeadingAndTrailingSpaces(value);
+
+    return value;
+}
+
+/**
+ * @brief Processes the line after the directive and value got extracted
+ *
+ * It can be the case that multiple directorives are in the same line.
+ * Therefore the line must still get read until all directives got processed.
+ *
+ * If the line still contains more then one directive, the processed directive gets skipped.
+ * If the line does not contain any more directives, the line gets cleared.
+ *
+ * @param line The current line
+ */
+void ConfigFileParser::processRemainingLine(std::string& line) const
+{
+    size_t semicolonIndex = line.find(';');
+
+    if (line.size() > semicolonIndex + 1) {
+        line = line.substr(semicolonIndex + 1);
+        removeLeadingAndTrailingSpaces(line);
+    } else
+        line.clear();
```

Comment:
I think you could just erase everything up to the first ';' you find, since you clear anyway.

So it would be `line.erase(0, line.find(';') + 1);`

Or a bit longer:
```
const size_t semicolonIndex = line.find(';');
line.erase(0, semicolonIndex + 1);
```

Reply:
Sums my approach up in fewer lines of code and makes it more readable. Nice!

Chose to use the short variant of your solution, as it remains easy to understand.